k2/csrc/weights.cc

```diff
@@ -0,0 +1,79 @@
+// k2/csrc/weights.cc
+
+// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/weights.h"
+
+#include <algorithm>
+#include <queue>
+#include <vector>
+
+#include "glog/logging.h"
+#include "k2/csrc/fsa.h"
+#include "k2/csrc/properties.h"
+
+namespace {
+using namespace k2;
+void CheckInput(const Fsa &fsa, const float *arc_weights) {
+  CHECK(IsValid(fsa));
+  CHECK(IsTopSorted(fsa));
+  CHECK_NOTNULL(arc_weights);
+}
+}  // namespace
+
+namespace k2 {
+
+void ComputeForwardMaxWeights(const Fsa &fsa, const float *arc_weights,
+                              float *state_weights) {
+  if (IsEmpty(fsa)) return;
+  CheckInput(fsa, arc_weights);
+  CHECK_NOTNULL(state_weights);
+
+  int32_t num_states = fsa.NumStates();
+  std::fill_n(state_weights, num_states, kNegativeInfinity);
+
+  const auto &arcs = fsa.arcs;
+  state_weights[0] = 0;
+  for (std::size_t i = 0; i != arcs.size(); ++i) {
+    const auto &arc = arcs[i];
+    auto src_weight = state_weights[arc.src_state];
+    auto &dest_weight = state_weights[arc.dest_state];
+    if (src_weight > kNegativeInfinity)
```

Comment:
I don't think this if stateent is needed.

Reply:
.. also this code seems to require topologically sorted input.  Double check that it's documented; and I think we should have a check that only runs in debug mode, that it's top sorted and acyclic.  E.g. within the loop, `CHECK(arc.dest_state > arc.src_state);`

Reply:
> I don't think this if stateent is needed.

It is needed here as we may update `dest_weight` to a value bigger than `kNegativeInfinity` (considering there's no standard `-inifity` in C++, I define it with `std::num_limits<float>::lowest`).

Reply:
> .. also this code seems to require topologically sorted input. Double check that it's documented; and I think we should have a check that only runs in debug mode, that it's top sorted and acyclic. E.g. within the loop, `CHECK(arc.dest_state > arc.src_state);`

OK. Will Update. Should I remove the check of `IsValid` as well?

Reply:
No don't remove the check of IsValid.
Regarding infinity, just use -std::numeric_limits<float>::infinity() wherever we need that.  It works fine, it's been used in Kaldi a lot.  And no need for the if.

Reply:
ok, seems compilers support `-std::numeric_limits::infinity()` well.